### ISSUE TITLE
Add -Werror flag

### ISSUE THREAD

**Issue by ndmitchell** [commercialhaskell/stack#3166]
Sometimes Stack gives warnings. These are easy to miss in the general output, and they don't fail a continuous integration build. It would be useful to have a flag that promoted these warnings into errors. Specifically the errors I'm thinking of are:

* Warnings from HPack.
* Warnings about missing modules not listed in other-modules.

**Comment by psibi** [comment]
There is a pedantic option for builds which does that IIRC. This is from the Stack help:

```
  --pedantic               Turn on -Wall and -Werror
```

**Comment by ndmitchell** [comment]
To concretely reproduce:

* Checkout Extra from https://github.com/ndmitchell/extra.git
* `stack init` to create a `stack.yaml` file
* Edit `extra.cabal` to delete the line mentioning `TestCustom`
* Run `stack test`

Observe that `stack` prints out:

```
Warning: The following modules should be added to exposed-modules or other-modules in C:\Neil\extra\extra.cabal:
    - In extra-test component:
        TestCustom
```

Turning on `--pedantic` has no impact at all. My understanding is that passes the appropriate error flags on GHC. In this case I'm after turning Stack/HPack warnings into errors. FWIW, I'd ideally like to turn on the errors to Stack/HPack independently of turning on `-Werror` to GHC.

**Comment by mgsloan** [comment]
Makes sense!  Probably don't want to have this apply to all warnings, there are some that you can't get rid of.

**Comment by ndmitchell** [comment]
@mgsloan - which warnings can't you get rid of it.

To my mind, asking Stack for errors and GHC for errors are not the same operations. I don't want `-Wall`, but I do want `-Werror`.

**Comment by mgsloan** [comment]
@ndmitchell One example is the warning that you should use `explicit-setup-deps` in your cabal file when using a custom Setup.hs.  It is not possible to resolve this warning while also retaining backwards compatibility with old cabal.

It may well make sense to have pedantic error out for stack warnings that ought to be resolvable (the majority).
